### PR TITLE
fix: cli command will not be executed if we add space after semicolon

### LINKED ISSUE #1121: CLI command will not be executed if we add space after semicolon
image
In sql_client, if we add space after semicolon and ENTER, command will not be executed

### PR DESCRIPTION
**Pull request by keyu813**
close #1121
image
only delete spaces after last semicolon if exist

### CODE REVIEW

**Comment by tobegit3hub** [comment]
Is this compatible with other databases? MySQL command client may execute event if it ends with spaces;

image

**Comment by keyu813** [comment]
> Is this compatible with other databases? MySQL command client may execute event if it ends with spaces;
> 
> image

Mysql support execute multiple commands (split by semicolon) at once, so "show databases; ;" will be explained as "show databases;" and " ;". Execution will get the result of "show databases" as well as Error(result of ";"). Openmldb's sql is a little different from Mysql, openmldb support execute a command at once temporarily. So I consider exit when there are multiple semicolons.

File: src/cmd/sql_cmd.h [hunk 1]
```diff
@@ -1469,6 +1468,18 @@ void Shell() {
                 continue;
             }
         }
+        if (buffer.find_last_of(';') != std::string::npos && buffer.back() != ';') {
+            std::string::size_type last_semicolon = buffer.find_last_of(';');
+            std::string::size_type buffer_size = buffer.size();
+            std::string after_semicolon_sql = buffer.substr(last_semicolon + 1, buffer_size);
+            boost::trim(after_semicolon_sql);
+            if (after_semicolon_sql.empty()) {
```
Comment (aceforeverd) [inline src/cmd/sql_cmd.h]: write tests to cover diff codes

**Comment by tobegit3hub** [comment]
@keyu813  We may support multiple commands in OpenMLDB soon. So "exit when there are multiple semicolons" seems not an good idea if we support multiple commands later.

File: src/cmd/sql_cmd.h [hunk 2]
```diff
@@ -1469,6 +1468,18 @@ void Shell() {
                 continue;
             }
         }
+        if (buffer.find_last_of(';') != std::string::npos && buffer.back() != ';') {
+            std::string::size_type last_semicolon = buffer.find_last_of(';');
+            std::string::size_type buffer_size = buffer.size();
+            std::string after_semicolon_sql = buffer.substr(last_semicolon + 1, buffer_size);
+            boost::trim(after_semicolon_sql);
```
Comment (aceforeverd) [inline src/cmd/sql_cmd.h]: avoid use boost, have a look at abseil https://github.com/abseil/abseil-cpp/blob/master/absl/strings/strip.h

**Comment by keyu813** [comment]
> @keyu813 We may support multiple commands in OpenMLDB soon. So "exit when there are multiple semicolons" seems not an good idea if we support multiple commands later.

ok, I will consider other solutions

File: src/cmd/sql_cmd.h [hunk 3]
```diff
@@ -1469,6 +1468,18 @@ void Shell() {
                 continue;
             }
         }
+        if (buffer.find_last_of(';') != std::string::npos && buffer.back() != ';') {
```
Comment (aceforeverd) [inline src/cmd/sql_cmd.h]: what's the use of semicolon ? 

so you'd like to cover the case like `show databases; show ` ? I checked mysql but they treat the second `show` as unterminated.

File: src/cmd/sql_cmd.h (hunk 1 above)
Comment (keyu813) [reply src/cmd/sql_cmd.h]: ok I'll add case

File: src/cmd/sql_cmd.h (hunk 2 above)
Comment (keyu813) [reply src/cmd/sql_cmd.h]: ok

File: src/cmd/sql_cmd.h (hunk 3 above)
Comment (keyu813) [reply src/cmd/sql_cmd.h]: I think mysql execute "show databases" and buffer "show" and wait for command, mysql support multiple commands. Openmldb support single command temporarily. So I will treat "show databases; show" as an illegal command. "Show databases; show databases;" will be consider as single command and parse failed. "Show database;      " will be executed.

**Review by vagetablechicken — CHANGES_REQUESTED** [review]
test `create procedure`

**Comment by keyu813** [comment]
> test `create procedure`

```
create procedure sp_auto_pro5(c1 string,c3 int,c7 timestamp)
        BEGIN
        SELECT * FROM t1 ; 
        END;
```
will failed (sql client can't read the whole sql because it will terminate when reading the line that ends with semicolon), but 
```
create procedure sp_auto_pro5(c1 string,c3 int,c7 timestamp)
        BEGIN
        SELECT * FROM t1 
        ;END;
```
can execute.

**Comment by aceforeverd** [comment]
How about the status now? @keyu813

**Review by dl239 — APPROVED** [review]

**Review by vagetablechicken — APPROVED** [review]

**Pull request merged**